statistics/semantic/src/main/java/com/spotify/heroic/statistics/semantic/SemanticMetricBackendReporter.java

```diff
@@ -118,19 +131,34 @@ public MetricBackend decorate(final MetricBackend backend) {
     @Override
     public DataInMemoryReporter newDataInMemoryReporter() {
         return new DataInMemoryReporter() {
-            private final AtomicLong dataRead = new AtomicLong(0);
-            private final AtomicLong dataInMemory = new AtomicLong(0);
-            private final AtomicLong rowsAccessed = new AtomicLong(0);
+            private final LongAccumulator dataRead = new LongAccumulator((x, y) -> x + y, 0L);
+            private final AtomicLong dataInMemory = new AtomicLong(0L);
+            private final LongAccumulator rowsAccessed = new LongAccumulator((x, y) -> x + y, 0L);
+            private final LongAccumulator maxDataInMemory = new LongAccumulator(Long::max, 0L);
+            private final Stopwatch queryWatch = Stopwatch.createStarted();
+            private final Map<Integer, ReadRangeStatistics> seriesDensity =
+                new ConcurrentHashMap<>();
 
             @Override
             public void reportRowsAccessed(final long n) {
-                rowsAccessed.addAndGet(n);
+                rowsAccessed.accumulate(n);
+            }
+
+            @Override
+            public void reportSliceRead(
+                final Series series, final DateRange range, final long n
+            ) {
+                int key = series.hashCode() + range.hashCode();
```

Comment:
Thanks for the advice!
I've been restructuring this part of the code. It's been a few days now since I touched the code so I don't have it fully in my head, but I believe my last experiment could manage without this data structure completely. So no more risk for collisions in that case. If it turns out that this is still needed, I'll go with your proposal for sure.